proof/refine/ARM/Tcb_R.thy

```diff
@@ -24,47 +24,262 @@ lemma activate_idle_thread_corres:
+
+lemma invs'_ko_at_valid_sc:
+  "ko_at' ko scp s \<Longrightarrow> invs' s \<Longrightarrow> valid_sched_context_size' ko \<and> valid_sched_context' ko s"
+  apply (drule invs_valid_objs')
+  by (fastforce simp: valid_objs'_def obj_at'_real_def ko_wp_at'_def valid_obj'_def projectKO_sc
+                      ran_def)
+
+lemma valid_sched_context'_scConsumed_update:
+  "valid_sched_context' ko s \<Longrightarrow> valid_sched_context' (scConsumed_update f ko) s"
+  by (clarsimp simp: valid_sched_context'_def)
+
+lemma schedContextUpdateConsumed_invs':
+  "schedContextUpdateConsumed scp \<lbrace>invs'\<rbrace>"
+  apply (simp add: schedContextUpdateConsumed_def)
+  apply (wpsimp wp: setSchedContext_invs')
+  apply (frule (1) invs'_ko_at_idle_sc_is_idle')
+  apply (frule (1) invs'_live_sc'_ex_nonz_cap_to')
+  apply (frule (1) invs'_ko_at_valid_sc)
+  by (fastforce elim: valid_sched_context'_scConsumed_update)
+
+lemma setConsumed_invs':
+  "setConsumed scp buffer \<lbrace>invs'\<rbrace>"
+  apply (simp add: setConsumed_def)
+  by (wpsimp wp: schedContextUpdateConsumed_invs'
+      | strengthen tcb_at_invs')+
+
+crunches setConsumed
+  for ksSchedulerAction[wp]: "\<lambda>s. P (ksSchedulerAction s)"
+  and tcb_at'[wp]: "\<lambda>s. Q (tcb_at' t s)"
+
+lemma schedContextCompleteYieldTo_invs':
+  "\<lbrace>invs' and sch_act_simple and tcb_at' thread\<rbrace>
+     schedContextCompleteYieldTo thread
+   \<lbrace>\<lambda>rv. invs'\<rbrace>"
+  unfolding schedContextCompleteYieldTo_def
+  by (wpsimp wp: schedContextCancelYieldTo_invs' setConsumed_invs'
+                 hoare_drop_imp hoare_vcg_if_lift2
+           simp: sch_act_simple_def)
+
+crunches schedContextCompleteYieldTo
+  for ksSchedulerAction[wp]: "\<lambda>s. P (ksSchedulerAction s)"
+  and tcb_at'[wp]: "\<lambda>s. Q (tcb_at' t s)"
+  and st_tcb_at'[wp]: "st_tcb_at' P t"
+
+lemma st_tcb_at'_valid_idle'_helper:
+  "st_tcb_at' P t s \<Longrightarrow> invs' s \<Longrightarrow> t = ksIdleThread s \<longrightarrow> P (IdleThreadState)"
+  by (clarsimp simp: invs'_def valid_state'_def valid_idle'_def pred_tcb_at'_def obj_at'_def
+                     idle_tcb'_def)
+
+lemma st_tcb_at'_valid_idle'_helper':
+  "invs' s \<Longrightarrow> \<forall>t. t = ksIdleThread s \<longrightarrow> st_tcb_at' idle' t s"
+  by (clarsimp simp: invs'_def valid_state'_def valid_idle'_def pred_tcb_at'_def obj_at'_def
+                     idle_tcb'_def)
+
+(* FIXME RT: move *)
+lemma maybeM_when:
+  "maybeM f x = when (x \<noteq> None) (f (the x))"
+  unfolding maybeM_def
+  by (clarsimp split: option.splits)
+
+lemma corres_assert_opt_assume_l'':
+  "corres_underlying sr nf nf' r P Q (f (the X)) g
+  \<Longrightarrow> corres_underlying sr nf nf' r (P and K (X \<noteq> None)) Q (assert_opt X >>= f) g"
+  by (force simp: corres_underlying_def assert_opt_def return_def bind_def fail_def)
+
+lemma scConsumed_update_corres:
+ "x = y \<Longrightarrow>
+  corres dc (sc_at scp) (ko_at' sc scp)
```

Comment:
Maybe investigate using `sc_at_cross` to remove the concrete guard? It'll need `pspace_aligned` and `pspace_distinct` but that's probably fine

Reply:
I think in this case we do need the explicit `ko_at` precondition.